File: laravel/app/Services/CoordinateResolver.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Exceptions\CoordinateResolutionException;
use App\Models\CoordinateMapping;
use App\ValueObjects\CoordinateData;

/**
 * Coordinate resolution service.
 *
 * Resolves coordinate keys through the three-layer registry system:
 * - Layer 1: Glossary (human metadata)
 * - Layer 2: Coordinate Address (unique identifier)
 * - Layer 3: Bitmask Policy (source of truth)
 *
 * All resolution is immutable and cached at the database level.
 */
final class CoordinateResolver
{
    private const COORDINATE_FORMAT = '/^[0-9a-f]{4}\.[0-9a-f]{4}@$/i';

    /**
     * Resolve a coordinate key to complete coordinate data.
     *
     * Fetches all three layers from the database and validates the format.
     *
     * @param string $coordinateKey The coordinate key to resolve (e.g., COORD_X101)
     * @return CoordinateData The resolved coordinate data
     * @throws CoordinateResolutionException
     */
    public function resolve(string $coordinateKey): CoordinateData
    {
        $mapping = $this->findMapping($coordinateKey);

        if (!$mapping) {
            throw CoordinateResolutionException::coordinateNotFound($coordinateKey);
        }

        $this->validateCoordinateAddress($mapping->coordinate_address);

        return CoordinateData::fromModel($mapping);
    }

    /**
     * Resolve multiple coordinates at once.
     *
     * @param string ...$coordinateKeys The coordinate keys to resolve
     * @return array<string, CoordinateData> Keyed by coordinate key
     * @throws CoordinateResolutionException If any coordinate is not found
     */
    public function resolveMany(string ...$coordinateKeys): array
    {
        $resolved = [];

        foreach ($coordinateKeys as $key) {
            $resolved[$key] = $this->resolve($key);
        }

        return $resolved;
    }

    /**
     * Try to resolve a coordinate, returning null if not found.
     *
     * @param string $coordinateKey The coordinate key to resolve
     * @return CoordinateData|null The resolved coordinate data or null
     */
    public function resolveOrNull(string $coordinateKey): ?CoordinateData
    {
        try {
            return $this->resolve($coordinateKey);
        } catch (CoordinateResolutionException) {
            return null;
        }
    }

    /**
     * Get only the bitmask policy for a coordinate (Layer 3).
     *
     * @param string $coordinateKey The coordinate key
     * @return int The bitmask policy value
     * @throws CoordinateResolutionException
     */
    public function resolveMask(string $coordinateKey): int
    {
        $mapping = $this->findMapping($coordinateKey);

        if (!$mapping) {
            throw CoordinateResolutionException::coordinateNotFound($coordinateKey);
        }

        return $mapping->bitmaskPolicy;
    }

    /**
     * Get only the glossary data for a coordinate (Layer 1).
     *
     * @param string $coordinateKey The coordinate key
     * @return array{
     *     label: string,
     *     description: string|null,
     *     seo_keywords: array<string>|null,
     *     schema_type: string|null
     * } The glossary data
     * @throws CoordinateResolutionException
     */
    public function resolveGlossary(string $coordinateKey): array
    {
        $mapping = $this->findMapping($coordinateKey);

        if (!$mapping) {
            throw CoordinateResolutionException::coordinateNotFound($coordinateKey);
        }

        return [
            'label' => $mapping->label ?? '',
            'description' => $mapping->description,
            'seo_keywords' => $mapping->seo_keywords,
            'schema_type' => $mapping->schema_type,
        ];
    }

    /**
     * Get only the coordinate address (Layer 2).
     *
     * @param string $coordinateKey The coordinate key
     * @return string The coordinate address
     * @throws CoordinateResolutionException
     */
    public function resolveAddress(string $coordinateKey): string
    {
        $mapping = $this->findMapping($coordinateKey);

        if (!$mapping) {
            throw CoordinateResolutionException::coordinateNotFound($coordinateKey);
        }

        return $mapping->coordinate_address;
    }

    /**
     * Check if a coordinate exists and is active.
     *
     * @param string $coordinateKey The coordinate key
     * @return bool True if coordinate exists and is active
     */
    public function exists(string $coordinateKey): bool
    {
        return $this->findMapping($coordinateKey) !== null;
    }

    /**
     * Validate coordinate address format.
     *
     * Format: XXXX.YYYY@ where X and Y are hexadecimal digits.
     *
     * @param string $address The coordinate address
     * @return void
     * @throws CoordinateResolutionException
     */
    public function validateCoordinateAddress(string $address): void
    {
        if (!preg_match(self::COORDINATE_FORMAT, $address)) {
            throw CoordinateResolutionException::invalidCoordinateFormat($address);
        }
    }

    /**
     * Find a mapping by coordinate key.
     *
     * Only returns active coordinates.
     *
     * @param string $coordinateKey The coordinate key
     * @return CoordinateMapping|null
     */
    private function findMapping(string $coordinateKey): ?CoordinateMapping
    {
        return CoordinateMapping::active()
            ->where('coordinate_key', $coordinateKey)
            ->first();
    }
}

File: python/clc/services/coordinate_resolver.py
import re
from typing import Optional, Dict, Any
from pydantic import BaseModel


class CoordinateData(BaseModel):
    coordinate_key: str
    label: str
    description: Optional[str] = None
    seo_keywords: Optional[list[str]] = None
    schema_type: Optional[str] = None
    coordinate_address: str
    bitmask_policy: int
    version: int
    is_active: bool

    def to_dict(self) -> Dict[str, Any]:
        return self.model_dump()

    def get_bitmask_hex(self) -> str:
        return hex(self.bitmask_policy)


class CoordinateResolver:
    COORDINATE_FORMAT = r"^[0-9a-f]{4}\.[0-9a-f]{4}@$"
    
    def __init__(self, registry_data: Dict[str, Any]):
        self.layer1 = registry_data.get("layer1_human_map", {})
        self.layer2 = registry_data.get("layer2_coordinate_registry", {})
        self.layer3 = registry_data.get("layer3_bitmask_core", {})

    def resolve(self, coordinate_key: str) -> CoordinateData:
        from clc.exceptions import CoordinateResolutionException

        if coordinate_key not in self.layer1:
            raise CoordinateResolutionException.coordinate_not_found(coordinate_key)

        glossary = self.layer1[coordinate_key]
        address = self.layer2.get(coordinate_key)
        mask = self.layer3.get(coordinate_key, 0)

        if not address:
            raise CoordinateResolutionException.layer_resolution_failed(2, coordinate_key)

        self._validate_coordinate_address(address)

        return CoordinateData(
            coordinate_key=coordinate_key,
            label=glossary.get("label", ""),
            description=glossary.get("description"),
            seo_keywords=glossary.get("seo_keywords"),
            schema_type=glossary.get("schema_type"),
            coordinate_address=address,
            bitmask_policy=mask,
            version=1,
            is_active=True,
        )

    def resolve_mask(self, coordinate_key: str) -> int:
        from clc.exceptions import CoordinateResolutionException

        if coordinate_key not in self.layer3:
            raise CoordinateResolutionException.coordinate_not_found(coordinate_key)

        return self.layer3[coordinate_key]

    def resolve_glossary(self, coordinate_key: str) -> Dict[str, Any]:
        from clc.exceptions import CoordinateResolutionException

        if coordinate_key not in self.layer1:
            raise CoordinateResolutionException.coordinate_not_found(coordinate_key)

        glossary = self.layer1[coordinate_key]
        return {
            "label": glossary.get("label", ""),
            "description": glossary.get("description"),
            "seo_keywords": glossary.get("seo_keywords"),
            "schema_type": glossary.get("schema_type"),
        }

    def resolve_address(self, coordinate_key: str) -> str:
        from clc.exceptions import CoordinateResolutionException

        if coordinate_key not in self.layer2:
            raise CoordinateResolutionException.coordinate_not_found(coordinate_key)

        return self.layer2[coordinate_key]

    def exists(self, coordinate_key: str) -> bool:
        return coordinate_key in self.layer1

    def _validate_coordinate_address(self, address: str) -> None:
        from clc.exceptions import CoordinateResolutionException

        if not re.match(self.COORDINATE_FORMAT, address):
            raise CoordinateResolutionException.invalid_coordinate_format(address)
